File: apps/api/AssetManagement.Api/Controllers/PeopleController.cs
using System.Globalization;
using AssetManagement.Api.Data;
using AssetManagement.Api.DTOs;
using AssetManagement.Api.Models;
using AssetManagement.Api.Services;
using CsvHelper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace AssetManagement.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/[controller]")]
public class PeopleController(AppDbContext db, IAuditService audit, ICurrentUserService currentUser) : ControllerBase
{
    [HttpGet("search")]
    public async Task<ActionResult<List<PersonSearchResult>>> Search(
        [FromQuery] string? q = null,
        [FromQuery] int limit = 5)
    {
        var query = db.People.Where(p => !p.IsArchived);

        if (!string.IsNullOrWhiteSpace(q))
            query = query.Where(p => EF.Functions.ILike(p.FullName, $"%{q}%"));

        var results = await query
            .OrderBy(p => p.FullName)
            .Take(limit)
            .Select(p => new PersonSearchResult(p.Id, p.FullName))
            .ToListAsync();

        return Ok(results);
    }

    [HttpGet]
    public async Task<ActionResult<PagedResponse<PersonDto>>> GetAll(
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 25,
        [FromQuery] string? search = null,
        [FromQuery] string sortBy = "fullname",
        [FromQuery] string sortDir = "asc")
    {
        page = Math.Max(1, page);
        pageSize = Math.Clamp(pageSize, 1, 100);

        var query = ApplySorting(BuildFilteredQuery(search), sortBy, sortDir);

        var totalCount = await query.CountAsync();

        var people = await query
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .Select(p => new PersonDto(
                p.Id, p.FullName, p.Email, p.Department, p.JobTitle,
                p.LocationId, p.Location != null ? p.Location.Name : null,
                p.IsArchived, p.CreatedAt, p.UpdatedAt))
            .ToListAsync();

        return Ok(new PagedResponse<PersonDto>(people, page, pageSize, totalCount));
    }

    [HttpGet("export")]
    public async Task<IActionResult> Export(
        [FromQuery] string? search = null,
        [FromQuery] string sortBy = "fullname",
        [FromQuery] string sortDir = "asc",
        [FromQuery] string? ids = null)
    {
        List<Person> people;
        if (!string.IsNullOrEmpty(ids))
        {
            var idList = ids.Split(',').Select(id => Guid.TryParse(id.Trim(), out var g) ? g : (Guid?)null).Where(g => g.HasValue).Select(g => g!.Value).ToList();
            people = await db.People.Include(p => p.Location).Where(p => !p.IsArchived && idList.Contains(p.Id)).ToListAsync();
        }
        else
        {
            var query = ApplySorting(BuildFilteredQuery(search), sortBy, sortDir);
            people = await query.Include(p => p.Location).ToListAsync();
        }

        var ms = new MemoryStream();
        await using var writer = new StreamWriter(ms, leaveOpen: true);
        await using var csv = new CsvWriter(writer, CultureInfo.InvariantCulture);

        csv.WriteField("FullName");
        csv.WriteField("Email");
        csv.WriteField("Department");
        csv.WriteField("JobTitle");
        csv.WriteField("Location");
        csv.WriteField("CreatedAt");
        csv.WriteField("UpdatedAt");
        await csv.NextRecordAsync();

        foreach (var p in people)
        {
            csv.WriteField(p.FullName);
            csv.WriteField(p.Email);
            csv.WriteField(p.Department);
            csv.WriteField(p.JobTitle);
            csv.WriteField(p.Location?.Name);
            csv.WriteField(p.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss"));
            csv.WriteField(p.UpdatedAt.ToString("yyyy-MM-dd HH:mm:ss"));
            await csv.NextRecordAsync();
        }

        await writer.FlushAsync();
        ms.Position = 0;

        return File(ms, "text/csv", "people-export.csv");
    }

    private IQueryable<Person> BuildFilteredQuery(string? search)
    {
        var query = db.People
            .Where(p => !p.IsArchived)
            .Include(p => p.Location)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            query = query.Where(p =>
                EF.Functions.ILike(p.FullName, $"%{search}%") ||
                (p.Email != null && EF.Functions.ILike(p.Email, $"%{search}%")));
        }

        return query;
    }

    private static IQueryable<Person> ApplySorting(IQueryable<Person> query, string sortBy, string sortDir)
    {
        var desc = string.Equals(sortDir, "desc", StringComparison.OrdinalIgnoreCase);
        return sortBy.ToLowerInvariant() switch
        {
            "email" => desc ? query.OrderByDescending(p => p.Email) : query.OrderBy(p => p.Email),
            "department" => desc ? query.OrderByDescending(p => p.Department) : query.OrderBy(p => p.Department),
            "jobtitle" => desc ? query.OrderByDescending(p => p.JobTitle) : query.OrderBy(p => p.JobTitle),
            "locationname" => desc ? query.OrderByDescending(p => p.Location!.Name) : query.OrderBy(p => p.Location!.Name),
            "createdat" => desc ? query.OrderByDescending(p => p.CreatedAt) : query.OrderBy(p => p.CreatedAt),
            _ => desc ? query.OrderByDescending(p => p.FullName) : query.OrderBy(p => p.FullName),
        };
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<PersonDto>> GetById(Guid id)
    {
        var person = await db.People
            .Include(p => p.Location)
            .FirstOrDefaultAsync(p => p.Id == id);

        if (person is null) return NotFound();

        return Ok(new PersonDto(
            person.Id, person.FullName, person.Email, person.Department, person.JobTitle,
            person.LocationId, person.Location?.Name,
            person.IsArchived, person.CreatedAt, person.UpdatedAt));
    }

    [HttpPost]
    public async Task<ActionResult<PersonDto>> Create(CreatePersonRequest request)
    {
        if (request.LocationId.HasValue)
        {
            var location = await db.Locations.FindAsync(request.LocationId.Value);
            if (location is null || location.IsArchived)
                return BadRequest("Invalid location ID.");
        }

        var person = new Person
        {
            Id = Guid.NewGuid(),
            FullName = request.FullName,
            Email = request.Email,
            Department = request.Department,
            JobTitle = request.JobTitle,
            LocationId = request.LocationId
        };

        db.People.Add(person);
        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Created",
            EntityType: "Person",
            EntityId: person.Id.ToString(),
            EntityName: person.FullName,
            Details: $"Created person \"{person.FullName}\"",
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName));

        // Reload with Location for response
        await db.Entry(person).Reference(p => p.Location).LoadAsync();

        var dto = new PersonDto(
            person.Id, person.FullName, person.Email, person.Department, person.JobTitle,
            person.LocationId, person.Location?.Name,
            person.IsArchived, person.CreatedAt, person.UpdatedAt);

        return CreatedAtAction(nameof(GetById), new { id = person.Id }, dto);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<PersonDto>> Update(Guid id, UpdatePersonRequest request)
    {
        var person = await db.People.Include(p => p.Location).FirstOrDefaultAsync(p => p.Id == id);
        if (person is null) return NotFound();

        if (request.LocationId.HasValue)
        {
            var location = await db.Locations.FindAsync(request.LocationId.Value);
            if (location is null || location.IsArchived)
                return BadRequest("Invalid location ID.");
        }

        var changes = new List<AuditChange>();
        Track(changes, "Full Name", person.FullName, request.FullName);
        Track(changes, "Email", person.Email, request.Email);
        Track(changes, "Department", person.Department, request.Department);
        Track(changes, "Job Title", person.JobTitle, request.JobTitle);

        if (request.LocationId != person.LocationId)
        {
            var oldName = person.Location?.Name;
            string? newName = null;
            if (request.LocationId is not null)
                newName = await db.Locations.Where(l => l.Id == request.LocationId).Select(l => l.Name).FirstAsync();
            changes.Add(new AuditChange("Location", oldName, newName));
        }

        person.FullName = request.FullName;
        person.Email = request.Email;
        person.Department = request.Department;
        person.JobTitle = request.JobTitle;
        person.LocationId = request.LocationId;
        person.UpdatedAt = DateTime.UtcNow;

        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Updated",
            EntityType: "Person",
            EntityId: person.Id.ToString(),
            EntityName: person.FullName,
            Details: $"Updated person \"{person.FullName}\"",
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName,
            Changes: changes));

        await db.Entry(person).Reference(p => p.Location).LoadAsync();

        return Ok(new PersonDto(
            person.Id, person.FullName, person.Email, person.Department, person.JobTitle,
            person.LocationId, person.Location?.Name,
            person.IsArchived, person.CreatedAt, person.UpdatedAt));
    }

    private static void Track(List<AuditChange> changes, string field, string? oldVal, string? newVal)
    {
        if (oldVal != newVal)
            changes.Add(new AuditChange(field, oldVal, newVal));
    }

    [HttpPost("bulk-archive")]
    public async Task<IActionResult> BulkArchive([FromBody] BulkArchiveRequest request)
    {
        int succeeded = 0, failed = 0;
        foreach (var id in request.Ids)
        {
            var person = await db.People.FindAsync(id);
            if (person is null || person.IsArchived) { failed++; continue; }

            person.IsArchived = true;
            person.UpdatedAt = DateTime.UtcNow;
            await db.SaveChangesAsync();

            await audit.LogAsync(new AuditEntry(
                Action: "Archived",
                EntityType: "Person",
                EntityId: person.Id.ToString(),
                EntityName: person.FullName,
                Details: $"Bulk archived person \"{person.FullName}\"",
                ActorId: currentUser.UserId,
                ActorName: currentUser.UserName));
            succeeded++;
        }
        return Ok(new BulkActionResponse(succeeded, failed));
    }

    [HttpGet("{id:guid}/history")]
    public async Task<ActionResult<List<PersonHistoryDto>>> GetHistory(
        Guid id,
        [FromQuery] int? limit = null)
    {
        var person = await db.People.FindAsync(id);
        if (person is null) return NotFound();

        var query = db.PersonHistory
            .Where(h => h.PersonId == id)
            .Include(h => h.PerformedByUser)
            .Include(h => h.Changes)
            .OrderByDescending(h => h.Timestamp)
            .AsQueryable();

        if (limit.HasValue)
            query = query.Take(limit.Value);

        var history = await query
            .Select(h => new PersonHistoryDto(
                h.Id,
                h.EventType.ToString(),
                h.Details,
                h.Timestamp,
                h.PerformedByUser != null ? h.PerformedByUser.DisplayName : null,
                h.Changes.Select(c => new PersonHistoryChangeDto(
                    c.FieldName, c.OldValue, c.NewValue
                )).ToList()))
            .ToListAsync();

        return Ok(history);
    }

    [HttpGet("{id:guid}/assets")]
    public async Task<ActionResult<List<AssignedAssetDto>>> GetAssignedAssets(Guid id)
    {
        var person = await db.People.FindAsync(id);
        if (person is null) return NotFound();

        var assets = await db.Assets
            .Where(a => a.AssignedPersonId == id && !a.IsArchived)
            .Include(a => a.AssetType)
            .Include(a => a.Location)
            .OrderBy(a => a.Name)
            .Select(a => new AssignedAssetDto(
                a.Id,
                a.Name,
                a.AssetTag,
                a.SerialNumber,
                a.Status.ToString(),
                a.AssetType!.Name,
                a.Location != null ? a.Location.Name : null))
            .ToListAsync();

        return Ok(assets);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Archive(Guid id)
    {
        var person = await db.People.FindAsync(id);
        if (person is null) return NotFound();

        person.IsArchived = true;
        person.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();

        await audit.LogAsync(new AuditEntry(
            Action: "Archived",
            EntityType: "Person",
            EntityId: person.Id.ToString(),
            EntityName: person.FullName,
            Details: $"Archived person \"{person.FullName}\"",
            ActorId: currentUser.UserId,
            ActorName: currentUser.UserName));

        return NoContent();
    }
}

File: apps/api-kt/src/main/kotlin/com/assetmanagement/api/controller/PeopleController.kt
package com.assetmanagement.api.controller

import com.assetmanagement.api.dto.*
import com.assetmanagement.api.model.Person
import com.assetmanagement.api.util.CsvUtils
import com.assetmanagement.api.repository.*
import com.assetmanagement.api.service.*
import com.opencsv.CSVWriter
import jakarta.persistence.criteria.Predicate
import jakarta.servlet.http.HttpServletResponse
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.http.ResponseEntity
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*
import java.io.OutputStreamWriter
import java.net.URI
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.*

@RestController
@RequestMapping("/api/v1/people")
class PeopleController(
    private val personRepository: PersonRepository,
    private val locationRepository: LocationRepository,
    private val assetRepository: AssetRepository,
    private val certificateRepository: CertificateRepository,
    private val applicationRepository: ApplicationRepository,
    private val personHistoryRepository: PersonHistoryRepository,
    private val auditService: AuditService,
    private val currentUserService: CurrentUserService
) {
    private val dateFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss").withZone(ZoneOffset.UTC)

    @GetMapping("/search")
    fun search(@RequestParam(required = false) q: String?, @RequestParam(defaultValue = "5") limit: Int): ResponseEntity<List<PersonSearchResult>> {
        val results = personRepository.search(q ?: "", PageRequest.of(0, limit))
            .map { PersonSearchResult(it.id, it.fullName) }
        return ResponseEntity.ok(results)
    }

    @GetMapping
    fun getAll(
        @RequestParam(defaultValue = "1") page: Int,
        @RequestParam(defaultValue = "25") pageSize: Int,
        @RequestParam(required = false) search: String?,
        @RequestParam(defaultValue = "fullname") sortBy: String,
        @RequestParam(defaultValue = "asc") sortDir: String
    ): ResponseEntity<PagedResponse<PersonDto>> {
        val p = maxOf(1, page)
        val ps = pageSize.coerceIn(1, 100)
        val spec = buildSpec(search)
        val sort = sortOf(sortBy, sortDir)
        val result = personRepository.findAll(spec, PageRequest.of(p - 1, ps, sort))
        val items = result.content.map { it.toDto() }
        return ResponseEntity.ok(PagedResponse(items, p, ps, result.totalElements))
    }

    @GetMapping("/export")
    fun export(
        @RequestParam(required = false) search: String?,
        @RequestParam(defaultValue = "fullname") sortBy: String,
        @RequestParam(defaultValue = "asc") sortDir: String,
        @RequestParam(required = false) ids: String?,
        response: HttpServletResponse
    ) {
        response.contentType = "text/csv"
        response.setHeader("Content-Disposition", "attachment; filename=people-export.csv")
        val people = if (!ids.isNullOrBlank()) {
            val idList = ids.split(",").mapNotNull { runCatching { UUID.fromString(it.trim()) }.getOrNull() }
            personRepository.findAllById(idList).filter { !it.isArchived }
        } else {
            personRepository.findAll(buildSpec(search), sortOf(sortBy, sortDir))
        }
        val writer = CSVWriter(OutputStreamWriter(response.outputStream))
        writer.writeNext(arrayOf("FullName", "Email", "Department", "JobTitle", "Location", "CreatedAt", "UpdatedAt"))
        people.forEach { p ->
            writer.writeNext(CsvUtils.sanitizeRow(arrayOf(p.fullName, p.email ?: "", p.department ?: "", p.jobTitle ?: "",
                p.location?.name ?: "", dateFormat.format(p.createdAt), dateFormat.format(p.updatedAt))))
        }
        writer.flush()
    }

    @GetMapping("/{id}")
    fun getById(@PathVariable id: UUID): ResponseEntity<PersonDto> {
        val person = personRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(person.toDto())
    }

    @GetMapping("/{id}/history")
    fun getHistory(@PathVariable id: UUID, @RequestParam(required = false) limit: Int?): ResponseEntity<Any> {
        if (!personRepository.existsById(id)) return ResponseEntity.notFound().build()
        val pageable = PageRequest.of(0, limit ?: 50)
        val history = personHistoryRepository.findByPersonIdOrderByTimestampDesc(id, pageable).map { h ->
            PersonHistoryDto(h.id, h.eventType.name, h.details, h.timestamp,
                h.performedByUser?.displayName,
                h.changes.map { c -> PersonHistoryChangeDto(c.fieldName, c.oldValue, c.newValue) })
        }
        return ResponseEntity.ok(history)
    }

    @GetMapping("/{id}/assets")
    fun getAssignedAssets(@PathVariable id: UUID): ResponseEntity<Any> {
        if (!personRepository.existsById(id)) return ResponseEntity.notFound().build()
        val spec = Specification<com.assetmanagement.api.model.Asset> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("assignedPersonId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }
        val assets = assetRepository.findAll(spec, Sort.by("name")).map { a ->
            AssignedAssetDto(a.id, a.name, a.serialNumber, a.status.name, a.assetType?.name ?: "", a.location?.name)
        }
        return ResponseEntity.ok(assets)
    }

    @PostMapping
    fun create(@RequestBody request: CreatePersonRequest): ResponseEntity<Any> {
        if (request.locationId != null) {
            val loc = locationRepository.findById(request.locationId).orElse(null)
            if (loc == null || loc.isArchived) return ResponseEntity.badRequest().body("Invalid location ID.")
        }
        val person = Person(fullName = request.fullName, email = request.email, department = request.department,
            jobTitle = request.jobTitle, locationId = request.locationId)
        personRepository.save(person)

        auditService.log(AuditEntry("Created", "Person", person.id.toString(), person.fullName,
            "Created person \"${person.fullName}\"", currentUserService.userId, currentUserService.userName))

        return ResponseEntity.created(URI("/api/v1/people/${person.id}")).body(personRepository.findById(person.id).get().toDto())
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: UUID, @RequestBody request: UpdatePersonRequest): ResponseEntity<Any> {
        val person = personRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        if (request.locationId != null) {
            val loc = locationRepository.findById(request.locationId).orElse(null)
            if (loc == null || loc.isArchived) return ResponseEntity.badRequest().body("Invalid location ID.")
        }

        val changes = mutableListOf<AuditChange>()
        if (person.fullName != request.fullName) changes.add(AuditChange("Full Name", person.fullName, request.fullName))
        if (person.email != request.email) changes.add(AuditChange("Email", person.email, request.email))
        if (person.department != request.department) changes.add(AuditChange("Department", person.department, request.department))
        if (person.jobTitle != request.jobTitle) changes.add(AuditChange("Job Title", person.jobTitle, request.jobTitle))

        person.fullName = request.fullName
        person.email = request.email
        person.department = request.department
        person.jobTitle = request.jobTitle
        person.locationId = request.locationId
        person.updatedAt = Instant.now()
        personRepository.save(person)

        auditService.log(AuditEntry("Updated", "Person", person.id.toString(), person.fullName,
            "Updated person \"${person.fullName}\"", currentUserService.userId, currentUserService.userName, changes))

        return ResponseEntity.ok(personRepository.findById(person.id).get().toDto())
    }

    @PostMapping("/bulk-archive")
    fun bulkArchive(@RequestBody request: BulkArchiveRequest): ResponseEntity<BulkActionResponse> {
        var succeeded = 0; var failed = 0
        request.ids.forEach { id ->
            val person = personRepository.findById(id).orElse(null)
            if (person == null || person.isArchived) { failed++; return@forEach }
            person.isArchived = true; person.updatedAt = Instant.now()
            personRepository.save(person)
            auditService.log(AuditEntry("Archived", "Person", person.id.toString(), person.fullName,
                "Bulk archived person \"${person.fullName}\"", currentUserService.userId, currentUserService.userName))
            succeeded++
        }
        return ResponseEntity.ok(BulkActionResponse(succeeded, failed))
    }

    @DeleteMapping("/{id}")
    fun archive(@PathVariable id: UUID): ResponseEntity<Any> {
        val person = personRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        person.isArchived = true; person.updatedAt = Instant.now()
        personRepository.save(person)
        auditService.log(AuditEntry("Archived", "Person", person.id.toString(), person.fullName,
            "Archived person \"${person.fullName}\"", currentUserService.userId, currentUserService.userName))
        return ResponseEntity.noContent().build()
    }

    // ── POST /check-duplicates ─────────────────────────────────────────
    @PostMapping("/check-duplicates")
    fun checkDuplicates(@RequestBody request: CheckPersonDuplicatesRequest): ResponseEntity<List<DuplicateCheckResult>> {
        val spec = Specification<Person> { root, _, cb ->
            val predicates = mutableListOf<Predicate>()
            predicates.add(cb.equal(root.get<Boolean>("isArchived"), false))

            if (request.excludeId != null)
                predicates.add(cb.notEqual(root.get<UUID>("id"), request.excludeId))

            val matchConditions = mutableListOf<Predicate>()

            if (!request.email.isNullOrBlank())
                matchConditions.add(cb.equal(cb.lower(root.get("email")), request.email.lowercase()))

            if (!request.fullName.isNullOrBlank())
                matchConditions.add(cb.like(cb.lower(root.get("fullName")), "%${request.fullName.lowercase()}%"))

            if (matchConditions.isEmpty())
                return@Specification cb.and(*predicates.toTypedArray(), cb.disjunction())

            predicates.add(cb.or(*matchConditions.toTypedArray()))
            cb.and(*predicates.toTypedArray())
        }

        val results = personRepository.findAll(spec, PageRequest.of(0, 5))
            .content.map { DuplicateCheckResult(it.id, it.fullName, "Email: ${it.email ?: "N/A"}") }

        return ResponseEntity.ok(results)
    }

    @GetMapping("/{id}/summary")
    fun getSummary(@PathVariable id: UUID): ResponseEntity<PersonSummaryDto> {
        if (!personRepository.existsById(id)) return ResponseEntity.notFound().build()

        val assetSpec = Specification<com.assetmanagement.api.model.Asset> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("assignedPersonId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }
        val certSpec = Specification<com.assetmanagement.api.model.Certificate> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("personId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }
        val appSpec = Specification<com.assetmanagement.api.model.Application> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("personId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }

        return ResponseEntity.ok(PersonSummaryDto(
            assetCount = assetRepository.count(assetSpec).toInt(),
            certificateCount = certificateRepository.count(certSpec).toInt(),
            applicationCount = applicationRepository.count(appSpec).toInt()
        ))
    }

    @GetMapping("/{id}/certificates")
    fun getAssignedCertificates(@PathVariable id: UUID): ResponseEntity<Any> {
        if (!personRepository.existsById(id)) return ResponseEntity.notFound().build()
        val spec = Specification<com.assetmanagement.api.model.Certificate> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("personId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }
        val certs = certificateRepository.findAll(spec, Sort.by("name")).map { c ->
            AssignedCertificateDto(c.id, c.name, c.certificateType?.name ?: "", c.status.name, c.expiryDate)
        }
        return ResponseEntity.ok(certs)
    }

    @GetMapping("/{id}/applications")
    fun getAssignedApplications(@PathVariable id: UUID): ResponseEntity<Any> {
        if (!personRepository.existsById(id)) return ResponseEntity.notFound().build()
        val spec = Specification<com.assetmanagement.api.model.Application> { root, _, cb ->
            cb.and(cb.equal(root.get<UUID>("personId"), id), cb.equal(root.get<Boolean>("isArchived"), false))
        }
        val apps = applicationRepository.findAll(spec, Sort.by("name")).map { a ->
            AssignedApplicationDto(a.id, a.name, a.applicationType?.name ?: "", a.status.name, a.licenceType?.name, a.expiryDate)
        }
        return ResponseEntity.ok(apps)
    }

    @PostMapping("/{id}/offboard")
    @Transactional
    fun offboard(@PathVariable id: UUID, @RequestBody request: OffboardRequest): ResponseEntity<Any> {
        val person = personRepository.findById(id).orElse(null) ?: return ResponseEntity.notFound().build()
        val results = mutableListOf<String>()

        for (action in request.actions) {
            when (action.entityType) {
                "Asset" -> {
                    val asset = assetRepository.findById(action.entityId).orElse(null) ?: continue
                    when (action.action) {
                        "free" -> {
                            asset.assignedPersonId = null
                            asset.status = com.assetmanagement.api.model.enums.AssetStatus.Available
                            asset.updatedAt = Instant.now()
                            assetRepository.save(asset)
                            auditService.log(AuditEntry("CheckedIn", "Asset", asset.id.toString(), asset.name,
                                "Checked in asset \"${asset.name}\" during offboarding of \"${person.fullName}\"",
                                currentUserService.userId, currentUserService.userName))
                            auditService.log(AuditEntry("AssetCheckedIn", "Person", person.id.toString(), person.fullName,
                                "Asset \"${asset.name}\" checked in during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Freed: ${asset.name}")
                        }
                        "transfer" -> {
                            val targetId = action.transferToPersonId ?: continue
                            val targetPerson = personRepository.findById(targetId).orElse(null) ?: continue
                            // Check in from source
                            asset.assignedPersonId = targetId
                            asset.status = com.assetmanagement.api.model.enums.AssetStatus.CheckedOut
                            asset.updatedAt = Instant.now()
                            assetRepository.save(asset)
                            auditService.log(AuditEntry("CheckedIn", "Asset", asset.id.toString(), asset.name,
                                "Checked in asset \"${asset.name}\" from \"${person.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            auditService.log(AuditEntry("AssetCheckedIn", "Person", person.id.toString(), person.fullName,
                                "Asset \"${asset.name}\" transferred to \"${targetPerson.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            // Check out to target
                            auditService.log(AuditEntry("CheckedOut", "Asset", asset.id.toString(), asset.name,
                                "Checked out asset \"${asset.name}\" to \"${targetPerson.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            auditService.log(AuditEntry("AssetCheckedOut", "Person", targetPerson.id.toString(), targetPerson.fullName,
                                "Asset \"${asset.name}\" transferred from \"${person.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Transferred: ${asset.name} → ${targetPerson.fullName}")
                        }
                        "keep" -> results.add("Kept: ${asset.name}")
                    }
                }
                "Certificate" -> {
                    val cert = certificateRepository.findById(action.entityId).orElse(null) ?: continue
                    when (action.action) {
                        "free" -> {
                            cert.personId = null
                            cert.updatedAt = Instant.now()
                            certificateRepository.save(cert)
                            auditService.log(AuditEntry("Updated", "Certificate", cert.id.toString(), cert.name,
                                "Unassigned certificate \"${cert.name}\" from \"${person.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Freed: ${cert.name}")
                        }
                        "transfer" -> {
                            val targetId = action.transferToPersonId ?: continue
                            val targetPerson = personRepository.findById(targetId).orElse(null) ?: continue
                            cert.personId = targetId
                            cert.updatedAt = Instant.now()
                            certificateRepository.save(cert)
                            auditService.log(AuditEntry("Updated", "Certificate", cert.id.toString(), cert.name,
                                "Transferred certificate \"${cert.name}\" from \"${person.fullName}\" to \"${targetPerson.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Transferred: ${cert.name} → ${targetPerson.fullName}")
                        }
                        "keep" -> results.add("Kept: ${cert.name}")
                    }
                }
                "Application" -> {
                    val app = applicationRepository.findById(action.entityId).orElse(null) ?: continue
                    when (action.action) {
                        "free" -> {
                            app.personId = null
                            app.updatedAt = Instant.now()
                            applicationRepository.save(app)
                            auditService.log(AuditEntry("Updated", "Application", app.id.toString(), app.name,
                                "Unassigned application \"${app.name}\" from \"${person.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Freed: ${app.name}")
                        }
                        "transfer" -> {
                            val targetId = action.transferToPersonId ?: continue
                            val targetPerson = personRepository.findById(targetId).orElse(null) ?: continue
                            app.personId = targetId
                            app.updatedAt = Instant.now()
                            applicationRepository.save(app)
                            auditService.log(AuditEntry("Updated", "Application", app.id.toString(), app.name,
                                "Transferred application \"${app.name}\" from \"${person.fullName}\" to \"${targetPerson.fullName}\" during offboarding",
                                currentUserService.userId, currentUserService.userName))
                            results.add("Transferred: ${app.name} → ${targetPerson.fullName}")
                        }
                        "keep" -> results.add("Kept: ${app.name}")
                    }
                }
            }
        }

        if (request.deactivatePerson) {
            person.isArchived = true
            person.updatedAt = Instant.now()
            personRepository.save(person)
            auditService.log(AuditEntry("Archived", "Person", person.id.toString(), person.fullName,
                "Archived person \"${person.fullName}\" during offboarding",
                currentUserService.userId, currentUserService.userName))
            results.add("Person archived")
        }

        return ResponseEntity.ok(OffboardResultDto(results))
    }

    private fun buildSpec(search: String?): Specification<Person> = Specification { root, _, cb ->
        val predicates = mutableListOf<Predicate>()
        predicates.add(cb.equal(root.get<Boolean>("isArchived"), false))
        if (!search.isNullOrBlank()) {
            val pattern = "%${search.lowercase()}%"
            predicates.add(cb.or(
                cb.like(cb.lower(root.get("fullName")), pattern),
                cb.like(cb.lower(root.get("email")), pattern)
            ))
        }
        cb.and(*predicates.toTypedArray())
    }

    private fun sortOf(sortBy: String, sortDir: String): Sort {
        val dir = if (sortDir.equals("desc", ignoreCase = true)) Sort.Direction.DESC else Sort.Direction.ASC
        val prop = when (sortBy.lowercase()) {
            "email" -> "email"; "department" -> "department"; "jobtitle" -> "jobTitle"
            "createdat" -> "createdAt"; else -> "fullName"
        }
        return Sort.by(dir, prop)
    }

    private fun Person.toDto() = PersonDto(id, fullName, email, department, jobTitle, locationId, location?.name, isArchived, createdAt, updatedAt)
}
